README.md
<p align="center"><a href="https://github.com/crazy-max/docker-rrdcached" target="_blank"><img height="128" src="https://raw.githubusercontent.com/crazy-max/docker-rrdcached/master/.github/docker-rrdcached.jpg"></a></p>

<p align="center">
  <a href="https://hub.docker.com/r/crazymax/rrdcached/tags?page=1&ordering=last_updated"><img src="https://img.shields.io/github/v/tag/crazy-max/docker-rrdcached?label=version&style=flat-square" alt="Latest Version"></a>
  <a href="https://github.com/crazy-max/docker-rrdcached/actions?workflow=build"><img src="https://img.shields.io/github/actions/workflow/status/crazy-max/docker-rrdcached/build.yml?branch=master&label=build&logo=github&style=flat-square" alt="Build Status"></a>
  <a href="https://hub.docker.com/r/crazymax/rrdcached/"><img src="https://img.shields.io/docker/stars/crazymax/rrdcached.svg?style=flat-square&logo=docker" alt="Docker Stars"></a>
  <a href="https://hub.docker.com/r/crazymax/rrdcached/"><img src="https://img.shields.io/docker/pulls/crazymax/rrdcached.svg?style=flat-square&logo=docker" alt="Docker Pulls"></a>
  <br /><a href="https://github.com/sponsors/crazy-max"><img src="https://img.shields.io/badge/sponsor-crazy--max-181717.svg?logo=github&style=flat-square" alt="Become a sponsor"></a>
  <a href="https://www.paypal.me/crazyws"><img src="https://img.shields.io/badge/donate-paypal-00457c.svg?logo=paypal&style=flat-square" alt="Donate Paypal"></a>
</p>

## About

[RRDcached](https://oss.oetiker.ch/rrdtool/doc/rrdcached.en.html) Docker image.

> **Note**
> 
> Want to be notified of new releases? Check out 🔔 [Diun (Docker Image Update Notifier)](https://github.com/crazy-max/diun)
> project!

___

* [Features](#features)
* [Build locally](#build-locally)
* [Image](#image)
* [Environment variables](#environment-variables)
* [Volumes](#volumes)
* [Ports](#ports)
* [Usage](#usage)
  * [Docker Compose](#docker-compose)
  * [Command line](#command-line)
* [Upgrade](#upgrade)
* [Contributing](#contributing)
* [License](#license)

## Features

* Run as non-root user
* Multi-platform image

## Build locally

```shell
git clone https://github.com/crazy-max/docker-rrdcached.git
cd docker-rrdcached

# Build image and output to docker (default)
docker buildx bake

# Build multi-platform image
docker buildx bake image-all
```

## Image

| Registry                                                                                         | Image                           |
|--------------------------------------------------------------------------------------------------|---------------------------------|
| [Docker Hub](https://hub.docker.com/r/crazymax/rrdcached/)                                            | `crazymax/rrdcached`                 |
| [GitHub Container Registry](https://github.com/users/crazy-max/packages/container/package/rrdcached)  | `ghcr.io/crazy-max/rrdcached`        |

Following platforms for this image are available:

```
$ docker run --rm mplatform/mquery crazymax/rrdcached:latest
Image: crazymax/rrdcached:latest
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64
   - linux/arm/v6
   - linux/arm/v7
   - linux/arm64
   - linux/386
   - linux/ppc64le
   - linux/s390x
```

## Environment variables

* `TZ` : Timezone assigned to the container (default `UTC`)
* `PUID` : Daemon user id (default `1000`)
* `PGID` : Daemon group id (default `1000`)
* `LOG_LEVEL` : Log level, called with `-V` (default `LOG_INFO`)
* `WRITE_TIMEOUT` : Data is written to disk every *X* seconds, called with `-w` (default `300`)
* `WRITE_JITTER` : Delay writing of each RRD for a random number of seconds in the range, called with `-z`
* `WRITE_THREADS` : Number of threads used for writing RRD files, called with `-t` (default `4`)
* `FLUSH_DEAD_DATA_INTERVAL` : Every *X* seconds the entire cache is searched for old values which are written to disk, called with `-f` (default `3600`)

> More info : https://github.com/oetiker/rrdtool-1.x/blob/master/doc/rrdcached.pod

## Volumes

* `/data/db` : Contains rrd database
* `/data/journal` :  Container rrd journal files

> :warning: Note that the volumes should be owned by the user/group with the specified `PUID` and `PGID`.
> If you don't give the volume correct permissions, the container may not start.

## Ports

* `42217` : RRDcached port

## Usage

### Docker Compose

Docker compose is the recommended way to run this image. You can use the following
[compose template](examples/compose/compose.yml), then run the container:

```bash
docker compose up -d
docker compose logs -f
```

### Command line

You can also use the following minimal command:

```bash
$ docker run -d -p 42217:42217 --name rrdcached \
  -e TZ="Europe/Paris" \
  -v $(pwd)/data/db:/data/db \
  -v $(pwd)/data/journal:/data/journal \
  crazymax/rrdcached
```

## Upgrade

Recreate the container whenever I push an update:

```bash
docker compose pull
docker compose up -d
```

## Contributing

Want to contribute? Awesome! The most basic way to show your support is to star
the project, or to raise issues. You can also support this project by [**becoming a sponsor on GitHub**](https://github.com/sponsors/crazy-max)
or by making a [PayPal donation](https://www.paypal.me/crazyws) to ensure this
journey continues indefinitely!

Thanks again for your support, it is much appreciated! :pray:

## License

MIT. See `LICENSE` for more details.


CHANGELOG.md
# Changelog

## 1.8.0-r5-r0 (2023/12/30)

* RRDtool cached 1.8.0-r5
* Alpine Linux 3.19

## 1.8.0-r2-r0 (2023/07/30)

* RRDtool cached 1.8.0-r2
* Alpine Linux 3.18

## 1.7.2-r5-r0 (2022/05/15)

* RRDtool cached 1.7.2-r5
* Alpine Linux 3.15

## 1.7.2-r4-r2 (2021/03/19)

* alpine-s6 3.14-2.2.0.3 (#30)
* Move to `docker/metadata-action`

## 1.7.2-r4-r1 (2021/03/18)

* Upstream Alpine update

## 1.7.2-r4-r0 (2021/02/17)

* RRDtool cached 1.7.2-r4
* `s6-overlay` 2.2.0.3
* Alpine Linux 3.13
* Fix daemon to listen on port 42217 (librenms/docker#124)
* Switch to buildx bake

## 1.7.2-r3-RC4 (2020/11/08)

* Switch to `s6-overlay` and log through `socklog-overlay`

## 1.7.2-r3-RC3 (2020/08/07)

* Set up healthcheck start period (#8)

## 1.7.2-r3-RC2 (2020/06/15)

* Alpine Linux 3.12

## 1.7.2-r3-RC1 (2020/04/04)

* RRDtool cached 1.7.2-r3
* Alpine Linux 3.11
* Switch to Open Container Specification labels as label-schema.org ones are deprecated

## 1.7.2-RC8 (2019/11/07)

* Bring back `PUID`/`PGID` vars

## 1.7.2-RC7 (2019/10/20)

* Fix `WRITE_JITTER`

## 1.7.2-RC6 (2019/10/17)

* Switch to GitHub Actions
* :warning: Stop publishing Docker image on Quay
* :warning: Run as non-root user
* Set timezone through tzdata

> :warning: **UPGRADE NOTES**
> As the Docker container now runs as a non-root user, you have to first stop the container and change permissions to volumes:
> ```
> docker compose stop
> chown -R 1000:1000 data/db data/journal
> docker compose pull
> docker compose up -d
> ```

## 1.7.2-RC5 (2019/09/14)

* Split db and journal volumes
* Add `PUID`/`PGID` vars

## 1.7.2-RC4 (2019/08/17)

* Fix journal issue

## 1.7.2-RC3 (2019/08/17)

* No need to chown `/data`
* Remove `PUID`/`PGID` vars

## 1.7.2-RC2 (2019/08/07)

* Add healthcheck

## 1.7.2-RC1 (2019/06/24)

* RRDtool cached 1.7.2
* Alpine Linux 3.10

## 1.7.0-RC2 (2019/01/31)

* Alpine Linux 3.9

## 1.7.0-RC1 (2018/12/03)

* Fix RRDtool cached version

## 20180706

* Initial version


build.sh
#!/usr/bin/env bash
set -e

PROJECT=rrdcached
BUILD_DATE=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
BUILD_TAG=docker_build
BUILD_WORKINGDIR=${BUILD_WORKINGDIR:-.}
DOCKERFILE=${DOCKERFILE:-Dockerfile}
VCS_REF=${TRAVIS_COMMIT::7}
RUNNING_TIMEOUT=120
RUNNING_LOG_CHECK="listening for connections"

PUSH_LATEST=${PUSH_LATEST:-true}
DOCKER_LOGIN=${DOCKER_LOGIN:-crazymax}
DOCKER_USERNAME=${DOCKER_USERNAME:-crazymax}
DOCKER_REPONAME=${DOCKER_REPONAME:-rrdcached}
QUAY_LOGIN=${QUAY_LOGIN:-crazymax}
QUAY_USERNAME=${QUAY_USERNAME:-crazymax}
QUAY_REPONAME=${QUAY_REPONAME:-rrdcached}

# Check local or travis
BRANCH=${TRAVIS_BRANCH:-local}
if [[ ${TRAVIS_PULL_REQUEST} == "true" ]]; then
  BRANCH=${TRAVIS_PULL_REQUEST_BRANCH}
fi
DOCKER_TAG=${BRANCH:-local}
if [[ "$BRANCH" == "master" ]]; then
  DOCKER_TAG=latest
elif [[ "$BRANCH" == "local" ]]; then
  BUILD_DATE=
  VERSION=local
fi

echo "PROJECT=${PROJECT}"
echo "VERSION=${VERSION}"
echo "BUILD_DATE=${BUILD_DATE}"
echo "BUILD_TAG=${BUILD_TAG}"
echo "BUILD_WORKINGDIR=${BUILD_WORKINGDIR}"
echo "DOCKERFILE=${DOCKERFILE}"
echo "VCS_REF=${VCS_REF}"
echo "PUSH_LATEST=${PUSH_LATEST}"
echo "DOCKER_LOGIN=${DOCKER_LOGIN}"
echo "DOCKER_USERNAME=${DOCKER_USERNAME}"
echo "DOCKER_REPONAME=${DOCKER_REPONAME}"
echo "QUAY_LOGIN=${QUAY_LOGIN}"
echo "QUAY_USERNAME=${QUAY_USERNAME}"
echo "QUAY_REPONAME=${QUAY_REPONAME}"
echo "TRAVIS_BRANCH=${TRAVIS_BRANCH}"
echo "TRAVIS_PULL_REQUEST=${TRAVIS_PULL_REQUEST}"
echo "BRANCH=${BRANCH}"
echo "DOCKER_TAG=${DOCKER_TAG}"
echo

# Build
echo "### Build"
docker build \
  --build-arg BUILD_DATE=${BUILD_DATE} \
  --build-arg VCS_REF=${VCS_REF} \
  --build-arg VERSION=${VERSION} \
  -t ${BUILD_TAG} -f ${DOCKERFILE} ${BUILD_WORKINGDIR}
echo

echo "### Test"
docker rm -f ${PROJECT} > /dev/null 2>&1 || true
docker run -d -p 42217:42217 --name ${PROJECT} ${BUILD_TAG}
echo

echo "### Waiting for ${PROJECT} to be up..."
TIMEOUT=$((SECONDS + RUNNING_TIMEOUT))
while read LOGLINE; do
  echo ${LOGLINE}
  if [[ ${LOGLINE} == *"${RUNNING_LOG_CHECK}"* ]]; then
    echo "Container up!"
    break
  fi
  if [[ $SECONDS -gt ${TIMEOUT} ]]; then
    >&2 echo "ERROR: Failed to run ${PROJECT} container"
    docker rm -f ${PROJECT} > /dev/null 2>&1 || true
    exit 1
  fi
done < <(docker logs -f ${PROJECT} 2>&1)
echo

CONTAINER_STATUS=$(docker container inspect --format "{{.State.Status}}" ${PROJECT})
if [[ ${CONTAINER_STATUS} != "running" ]]; then
  >&2 echo "ERROR: Container ${PROJECT} returned status '$CONTAINER_STATUS'"
  docker rm -f ${PROJECT} > /dev/null 2>&1 || true
  exit 1
fi
docker rm -f ${PROJECT} > /dev/null 2>&1 || true

if [ "${VERSION}" == "local" -o "${TRAVIS_PULL_REQUEST}" == "true" ]; then
  echo "INFO: This is a PR or a local build, skipping push..."
  exit 0
fi
if [[ ! -z ${DOCKER_PASSWORD} ]]; then
  echo "### Push to Docker Hub..."
  echo "$DOCKER_PASSWORD" | docker login --username "$DOCKER_LOGIN" --password-stdin > /dev/null 2>&1
  if [ "${DOCKER_TAG}" == "latest" -a "${PUSH_LATEST}" == "true" ]; then
    docker tag ${BUILD_TAG} ${DOCKER_USERNAME}/${DOCKER_REPONAME}:${DOCKER_TAG}
  fi
  if [[ "${VERSION}" != "latest" ]]; then
    docker tag ${BUILD_TAG} ${DOCKER_USERNAME}/${DOCKER_REPONAME}:${VERSION}
  fi
  docker push ${DOCKER_USERNAME}/${DOCKER_REPONAME}
  echo
fi
if [[ ! -z ${QUAY_PASSWORD} ]]; then
  echo "### Push to Quay..."
  echo "$QUAY_PASSWORD" | docker login quay.io --username "$QUAY_LOGIN" --password-stdin > /dev/null 2>&1
  if [ "${DOCKER_TAG}" == "latest" -a "${PUSH_LATEST}" == "true" ]; then
    docker tag ${BUILD_TAG} quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}:${DOCKER_TAG}
  fi
  if [[ "${VERSION}" != "latest" ]]; then
    docker tag ${BUILD_TAG} quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}:${VERSION}
  fi
  docker push quay.io/${QUAY_USERNAME}/${QUAY_REPONAME}
  echo
fi


entrypoint.sh
#!/bin/sh

if [ -n "${PGID}" ] && [ "${PGID}" != "$(id -g rrdcached)" ]; then
  echo "Switching to PGID ${PGID}..."
  sed -i -e "s/^rrdcached:\([^:]*\):[0-9]*/rrdcached:\1:${PGID}/" /etc/group
  sed -i -e "s/^rrdcached:\([^:]*\):\([0-9]*\):[0-9]*/rrdcached:\1:\2:${PGID}/" /etc/passwd
fi
if [ -n "${PUID}" ] && [ "${PUID}" != "$(id -u rrdcached)" ]; then
  echo "Switching to PUID ${PUID}..."
  sed -i -e "s/^rrdcached:\([^:]*\):[0-9]*:\([0-9]*\)/rrdcached:\1:${PUID}:\2/" /etc/passwd
fi

echo "Creating configuration..."
cat > /etc/rrdcached/rrdcached.conf <<EOL
LOG_LEVEL=${LOG_LEVEL:-LOG_INFO}
WRITE_TIMEOUT=${WRITE_TIMEOUT:-300}
WRITE_JITTER=${WRITE_JITTER:-0}
WRITE_THREADS=${WRITE_THREADS:-4}
FLUSH_DEAD_DATA_INTERVAL=${FLUSH_DEAD_DATA_INTERVAL:-3600}
EOL

echo "Fixing perms..."
chown -R rrdcached:rrdcached /etc/rrdcached /var/run/rrdcached
chown rrdcached:rrdcached /data /data/db /data/journal

exec su-exec rrdcached:rrdcached "$@"


rootfs/etc/cont-init.d/04-svc-main.sh
#!/usr/bin/with-contenv bash

LOG_LEVEL=${LOG_LEVEL:-LOG_INFO}
WRITE_TIMEOUT=${WRITE_TIMEOUT:-300}
WRITE_JITTER=${WRITE_JITTER:-0}
WRITE_THREADS=${WRITE_THREADS:-4}
FLUSH_DEAD_DATA_INTERVAL=${FLUSH_DEAD_DATA_INTERVAL:-3600}

JITTER=""
if [ "${WRITE_JITTER}" -gt "0" ]; then
  JITTER="-z ${WRITE_JITTER}"
fi

# https://oss.oetiker.ch/rrdtool/doc/rrdcached.en.html
mkdir -p /etc/services.d/rrdcached
cat > /etc/services.d/rrdcached/run <<EOL
#!/usr/bin/execlineb -P
with-contenv
/usr/sbin/rrdcached \
  -g \
  -F \
  -B \
  -R \
  -l :42217 \
  -p /var/run/rrdcached/rrdcached.pid \
  -b /data/db \
  -j /data/journal \
  -U rrdcached \
  -G rrdcached \
  -w "${WRITE_TIMEOUT}" \
  ${JITTER} \
  -f "${FLUSH_DEAD_DATA_INTERVAL}" \
  -t "${WRITE_THREADS}" \
  -V "${LOG_LEVEL}"
EOL
chmod +x /etc/services.d/rrdcached/run
